eqt/ui/UIFormWidget.py

```diff
@@ -263,39 +374,78 @@ def getWidgetState(self, widget, role=None):
+        
+        Parameters
+        -------------
+        name_key: str
+            Format: name or name_field or name_label
+        '''
+        if name_key.endswith('_field'):
+            name = name_key.removesuffix('_field')
+            role = 'field'
+        elif name_key.endswith('_label'):
+            name = name_key.removesuffix('_label')
+            role = 'label'
+        else:
+            name = name_key
+            role = 'field'
+        return name, role
+
+    def _getNameAndRoleFromWidget(self, widget):
+        '''       
+        Given a widget, finds it in the widget dictionary and returns its name and role. 
+        
+        Parameters
+        -------------
+        widget: qwidget
+        '''
+        for key, value in self.widgets.items():
+            if value == widget:
+                name, role = self._getNameAndRoleFromNameKey(key)
+        return name, role
+
+    def applyWidgetState(self, name_key, state, role=None):
         '''
-        Applies the given `state` to the widget with the given `name`.
+        Applies the given `state` to the widget associated with `name_key`.
 
         Parameters
         ----------
-        name: str
-            The name of the widget to apply the state to.
+        name_key: str
+            The key of the widget to apply the state to. Format: name or f'{name}_field' or f'{name}_label'
         role: str, optional, default None, values: 'label', 'field', None.
             The role of the widget to apply the state to (only if `widget` is a `str`).
-            If unspecified, the widget is chosen based on `name`.
+            If None, the widget is chosen based on the format of name_key.
         state: dict
-            Format: {'value': str | bool | int, 'enabled': bool, 'visible': bool},
-            e.g. {'value': 1, 'enabled': True, 'visible': True}.
+            Format: {'value': str | bool | int, 'enabled': bool, 'visible': bool, 'widget_number' : int},
+                    or {'value': str | bool | int, 'enabled': bool, 'visible': bool}
+                    e.g. {'value': 1, 'enabled': True, 'visible': True}.
         '''
         if role is not None:
             if role not in ['label', 'field']:
-                raise ValueError('role must be either "label", "field" or None')
-            name = name + '_' + role
+                raise ValueError('Role must be either "label", "field" or None')
+            name = name_key
+        else:
+            name, role = self._getNameAndRoleFromNameKey(name_key)
+        name_role = name + '_' + role
 
+        #retrieve widget
         try:
-            widget = self.widgets[name]
+            if name_role in self.widgets.keys():
+                widget = self.widgets[name_role]
+            elif hasattr(self, 'removed_widgets_dictionary'):
+                if name_role in self.removed_widgets_dictionary.keys():
+                    widget = self.removed_widgets_dictionary[name_role]
```

Comment:
Why would you want to return the widget dictionary from the removed widgets in this method?

Wouldn't it make more sense to get that info from a special method?

Reply:
The "apply state" method will get the widget either from the current widgets or from the removed widgets dictionary. The items in the state are applied to the widget here. Only in the "apply widget states (plural)" method the widget is restored in the form, not here.

Reply:
After discussing with @paskino on 12/01/24 we decided to move this procedure from the ApplyWidgetStates method.